src/todoApp.js

```diff
@@ -0,0 +1,68 @@
+import { generateKey } from './util.js';
+
+const $ulist = document.getElementById('todo-list');
+
+const todoApp = (todoInput, todoList, todoStatus) => {
+  let _todoItems = new Map();
+
+  const _filterStatusPredicate = {
+    //TODO sync with todoStatus.js/filters
+    all: () => true,
+    completed: ({ status }) => status === 'completed',
+    active: ({ status }) => status === '',
+  };
+
+  const _createTodoItem = (content, status) => {
+    const todoItem = {
+      index: generateKey(),
+      content,
+      status,
+    };
+    return todoItem;
+  };
+
+  const addTodoItem = (content, status = '') => {
+    const todoItem = _createTodoItem(content, status);
+    _todoItems.set(todoItem.index, todoItem);
+
+    todoListHandler.addItem(todoItem);
+    todoStatusHandler.updateCount();
+  };
+
+  const removeTodoItem = ({ index }) => {
+    _todoItems.delete(index);
+    todoStatusHandler.updateCount();
+  };
+
+  const updateTodoItem = ({ index, content, status }) => {
+    const todoItem = _todoItems.get(index);
+    todoItem.content = content;
+    todoItem.status = status ?? '';
+
+    _todoItems.set(todoItem.index, todoItem);
+  };
+
+  const setFilter = (filterType) => {
+    const filteredItems = Array.from(_todoItems.values()).filter(
+      _filterStatusPredicate[filterType]
+    );
```

Comment:
`가급적 원본을 변경하지 않는 Array메서드를 사용하려고는 했는데, 이와같이 state를 공유할 때 원본 객체의 변경을 막을 방법이나 우회 팁이 있을지 궁금합니다. freeze 하기엔 불변객체가 아니고.. 메서드에서의 read-only를 보장하는 방법이 있을까요? 매번 복사본을 생성??`

위 질문에 대해서는, 현재 프로퍼티의 밸류가 모두 원시타입인 string이기때문에 아래와같이 수정한다면 불변성을 유지할 수 있습니다.

물론 프로퍼티 밸류가 원시타입이 아니라면 보장되지 않겠지만요..

제가생각한 방법은 수지님이 말씀하신 매번 복사본을 생성하는 방법 인것 같네요.

```suggestion
  const setFilter = (filterType) => {
       const filteredItems = Array.from(_todoItems.values())
      .map((item) => ({ ...item }))
      .filter(_filterStatusPredicate[filterType]);
```

Reply:
프레임워크를 쓸때 참고하면 좋을 부분이겠네요..! 매번 복사본을 생성하는게 과한(?) 작업일까 고민되었는데, 답변 감사드립니다.